src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/JsonRequestHelpers.cs

```diff
@@ -265,13 +276,39 @@ public static async Task WriteResponseMessage(HttpResponse response, Encoding en
             GrpcServerLog.ReceivedMessage(serverCallContext.Logger);
             return (TRequest)requestMessage;
         }
+        catch (JsonException ex)
+        {
+            GrpcServerLog.ErrorReadingMessage(serverCallContext.Logger, ex);
+            throw new RpcException(new Status(StatusCode.InvalidArgument, "Request JSON payload is not correctly formatted."));
+        }
         catch (Exception ex)
         {
             GrpcServerLog.ErrorReadingMessage(serverCallContext.Logger, ex);
-            throw;
+            throw new RpcException(new Status(StatusCode.InvalidArgument, ex.Message));
         }
     }
 
+    private static async ValueTask<IMessage> ReadHttpBodyAsync(JsonTranscodingServerCallContext serverCallContext)
+    {
+        var ms = new MemoryStream();
```

Comment:
Should we back this with a `ArrayPool<byte>.Shared` in most cases like we do in [FileBufferingReadStream](https://github.com/dotnet/aspnetcore/blob/f041d49f3320fd0fb70babe35a861a9035e48e7e/src/Http/WebUtilities/src/FileBufferingReadStream.cs#L85-L95)? I figure this should help for smaller request bodies. At the very least, the initial size should be the Content-Length if it's provided.

Another option might be to read this into an unbounded `Pipe` and then copy the data into a pooled array when we're done, but I think continuing to use `MemoryStream` would be simpler.

Reply:
Should we just use `FileBufferingReadStream` under the cover for not-already-buffered `HttpBody` requests? I'm sure it'd slow things down for requests that end up buffering to a file (I think this is anything over 1MB by default), but we wouldn't blow up memory for larger bodies as much. What do you think @Tratcher?

Reply:
Yes, prefer FileBufferingReadStream  over MemoryStream unless you're certain the body is small.

Reply:
The only thing is the point is kinda moot if we're just going to turn it into an array anyway. It'd just save memory during the upload itself. But maybe that's still enough to make it worthwhile.

Reply:
FileBufferingReadStream  is pretty terrible for performance. We need to fix the block size problems.